### ISSUE TITLE
Missing xdgenvpy 

### ISSUE THREAD

**Issue by joakimkjellsson** [esm-tools/esm_tools#836]
Good morning
I'm still using ESM-Tools branch prep_release_awicm3_geomar (which is at least a year old) on HLRN-Emmy (Gottingen) but now I'm feeling brave enough to checkout "geomar_dev" (which is much newer). 

Doing "git checkout geomar_dev" worked fine. To be safe, I also removed my ~/.local folder and ran the "./install.sh" script anew. 
But when I try to submit a job, I get this error
```bash
(base) glogin7:~/esm/esm_tools $ esm_runscripts runscripts/focioifs/focioifs-piCtl-restart-glogin.yaml -e FOCI_GJK008 -c 
Traceback (most recent call last):
  File "/home/shkjocke/.local/bin/esm_runscripts", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/sw/tools/python/anaconda3/2019.03/skl/lib/python3.7/site-packages/pkg_resources/__init__.py", line 3191, in <module>
    @_call_aside
  File "/sw/tools/python/anaconda3/2019.03/skl/lib/python3.7/site-packages/pkg_resources/__init__.py", line 3175, in _call_aside
    f(*args, **kwargs)
  File "/sw/tools/python/anaconda3/2019.03/skl/lib/python3.7/site-packages/pkg_resources/__init__.py", line 3204, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/sw/tools/python/anaconda3/2019.03/skl/lib/python3.7/site-packages/pkg_resources/__init__.py", line 583, in _build_master
    ws.require(__requires__)
  File "/sw/tools/python/anaconda3/2019.03/skl/lib/python3.7/site-packages/pkg_resources/__init__.py", line 900, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/sw/tools/python/anaconda3/2019.03/skl/lib/python3.7/site-packages/pkg_resources/__init__.py", line 786, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'xdgenvpy' distribution was not found and is required by esm-tools
```
I can't find the "xdgenvpy" in my ~/.local folder so it seems that it wasn't installed. This is weird since it's in the "requirements" list in setup.py. 
I should mention that I've also loaded "module load git anaconda3" before doing any of this, and that the "geomar_dev" branch works fine on HLRN-Lise (berlin). 

Has anyone else managed to run recent ESM-Tools versions on HLRN-Emmy? Is there a way to force xdgenvpy to install? 

Best
Joakim

**Comment by seb-wahl** [comment]
Good morning,
I'd first do a complete uninstall as described here: https://esm-tools.readthedocs.io/en/latest/installation.html#uninstall-esm-tools
Does that do the trick?

**Comment by joakimkjellsson** [comment]
Thanks @seb-wahl 

It sort of worked. 
This seems to be a problem for me: 
```bash
(base) glogin6:~ $ pip freeze | grep esm | xargs pip uninstall -y
WARNING: pip is being invoked by an old script wrapper. This will fail in a future version of pip.
Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.
WARNING: pip is being invoked by an old script wrapper. This will fail in a future version of pip.
Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.

Usage:   
  pip uninstall [options] <package> ...
  pip uninstall [options] -r <requirements file> ...

no such option: -e
```

I don't know if this command actually did something or not. 
I ran the "./install.sh" afterwards, and now xdgenvpy is installed, so that's good. 

But the install script ended with
```bash
Requirement already satisfied: cftime in /sw/tools/python/anaconda3/2019.03/skl/lib/python3.7/site-packages (from netcdf4->gfw-creator->esm-tools==6.13.5) (1.0.3.4)
Requirement already satisfied: pycparser in /sw/tools/python/anaconda3/2019.03/skl/lib/python3.7/site-packages (from cffi>=1.4.1->pynacl>=1.4.0->PyGithub->esm-tools==6.13.5) (2.19)
Building wheels for collected packages: emoji
  Building wheel for emoji (setup.py) ... done
  Created wheel for emoji: filename=emoji-2.1.0-py3-none-any.whl size=212372 sha256=039af3f22fc719b30351915f839d295e98fc1a58d6962ce8e96542d17ce86006
  Stored in directory: /home/shkjocke/.cache/pip/wheels/77/75/99/51c2a119f4cfd3af7b49cc57e4f737bed7e40b348a85d82804
Successfully built emoji
Installing collected packages: xdgenvpy, semver, numpy, emoji, pandas, esm-tools
  Running setup.py develop for esm-tools
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tables 3.5.1 requires mock>=2.0, which is not installed.
datashader 0.7.0 requires testpath<0.4, but you have testpath 0.4.2 which is incompatible.
Successfully installed emoji-2.1.0 esm-tools numpy-1.21.6 pandas-1.3.5 semver-2.13.0 xdgenvpy-2.3.5
(base) glogin6:~/esm/esm_tools $
```

Then I re-ran the uninstall and install commands again, and this time it ended with
```bash
Requirement already satisfied: pycparser in /sw/tools/python/anaconda3/2019.03/skl/lib/python3.7/site-packages (from cffi>=1.4.1->pynacl>=1.4.0->PyGithub->esm-tools==6.13.5) (2.19)
Installing collected packages: esm-tools
  Running setup.py develop for esm-tools
Successfully installed esm-tools
```

But when I try to use esm_tools I get this:
```bash
(base) glogin6:~/esm/esm_tools $ esm_runscripts runscripts/focioifs/focioifs-piCtl-restart-glogin.yaml -e FOCI_GJK008 -c 
/home/shkjocke/.local/lib/python3.7/site-packages/pandas/compat/_optional.py:138: UserWarning: Pandas requires version '2.7.0' or newer of 'numexpr' (version '2.6.9' currently installed).
  warnings.warn(msg, UserWarning)
/sw/tools/python/anaconda3/2019.03/skl/lib/python3.7/site-packages/dask/config.py:168: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  data = yaml.load(f.read()) or {}
Traceback (most recent call last):
  File "/home/shkjocke/.local/bin/esm_runscripts", line 11, in <module>
    load_entry_point('esm-tools', 'console_scripts', 'esm_runscripts')()
  File "/home/shkjocke/esm/esm_tools/src/esm_runscripts/cli.py", line 274, in main
    Setup = SimulationSetup(command_line_config)
  File "/home/shkjocke/esm/esm_tools/src/esm_runscripts/sim_objects.py", line 42, in __init__
    self.config = prepare.run_job(self.config)
  File "/home/shkjocke/esm/esm_tools/src/esm_runscripts/prepare.py", line 17, in run_job
    helpers.evaluate(config, "prepare", "prepare_recipe")
  File "/home/shkjocke/esm/esm_tools/src/esm_runscripts/helpers.py", line 65, in evaluate
    plugins_bare, framework_recipe, need_to_parse_plugins
  File "/home/shkjocke/esm/esm_tools/src/esm_plugin_manager/esm_plugin_manager.py", line 59, in read_plugin_information
    attach_installed_plugins_to_all(plugins)
  File "/home/shkjocke/esm/esm_tools/src/esm_plugin_manager/esm_plugin_manager.py", line 74, in attach_installed_plugins_to_all
    plugins.update(find_installed_plugins())
  File "/home/shkjocke/esm/esm_tools/src/esm_plugin_manager/esm_plugin_manager.py", line 68, in find_installed_plugins
    for entry_point in pkg_resources.iter_entry_points("esm_tools.plugins")
  File "/home/shkjocke/esm/esm_tools/src/esm_plugin_manager/esm_plugin_manager.py", line 68, in <dictcomp>
    for entry_point in pkg_resources.iter_entry_points("esm_tools.plugins")
  File "/sw/tools/python/anaconda3/2019.03/skl/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2411, in load
    return self.resolve()
  File "/sw/tools/python/anaconda3/2019.03/skl/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2417, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/home/shkjocke/.local/lib/python3.7/site-packages/gfw_creator/__init__.py", line 9, in <module>
    from .gfw_creator import *
  File "/home/shkjocke/.local/lib/python3.7/site-packages/gfw_creator/gfw_creator.py", line 6, in <module>
    import xarray as xr
  File "/sw/tools/python/anaconda3/2019.03/skl/lib/python3.7/site-packages/xarray/__init__.py", line 9, in <module>
    from .core.combine import concat, auto_combine
  File "/sw/tools/python/anaconda3/2019.03/skl/lib/python3.7/site-packages/xarray/core/combine.py", line 9, in <module>
    from .merge import merge
  File "/sw/tools/python/anaconda3/2019.03/skl/lib/python3.7/site-packages/xarray/core/merge.py", line 17, in <module>
    PANDAS_TYPES = (pd.Series, pd.DataFrame, pd.Panel)
  File "/home/shkjocke/.local/lib/python3.7/site-packages/pandas/__init__.py", line 244, in __getattr__
    raise AttributeError(f"module 'pandas' has no attribute '{name}'")
AttributeError: module 'pandas' has no attribute 'Panel'
```

So it seems to have installed esm_tools, but there are a bunch of incompatibilities with the versions of packages. 
And since pip warned me that 
```bash
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
``` 
I assume that's the problem. 
Maybe the problem is that the command 
```bash
pip freeze | grep esm | xargs pip uninstall -y
```
only removes packages starting with "esm". I tried to run
```bash
pip freeze | xargs pip uninstall -y
```
but then it tries to remove both local and non-local installs, which I'm not allowed to do on HLRN (thank god!). 

How do I solve this? 

Many thanks
/J

**Comment by joakimkjellsson** [comment]
Update: I made a new conda environment and ran ./install.sh 

Now the ./install script runs fine, but when I try to make an experiment, I get this: 

```bash
(esm) esm_runscripts runscripts/focioifs/focioifs-piCtl-restart-glogin.yaml -e FOCI_GJK008 -c
First year, checking if we need to compile...

================================================================================
STARTING SIMULATION JOB!
Experiment ID = FOCI_GJK008
Setup = focioifs
This setup consists of:
- oifs
- nemo
- xios
- oasis3mct
- rnfmap
Experiment is installed in:
       /home/shkjocke/esm/esm-experiments//FOCI_GJK008
================================================================================

WARNING: File not found: /home/shkjocke/esm/esm-experiments/FOCI_GJK006/restart/oifs/18500101/rcf
Traceback (most recent call last):
  File "/home/shkjocke/.local/bin/esm_runscripts", line 33, in <module>
    sys.exit(load_entry_point('esm-tools', 'console_scripts', 'esm_runscripts')())
  File "/home/shkjocke/esm/esm_tools/src/esm_runscripts/cli.py", line 278, in main
    Setup()
  File "/home/shkjocke/esm/esm_tools/src/esm_runscripts/sim_objects.py", line 69, in __call__
    self.prepcompute()
  File "/home/shkjocke/esm/esm_tools/src/esm_runscripts/sim_objects.py", line 161, in prepcompute
    self.config = prepcompute.run_job(self.config)
  File "/home/shkjocke/esm/esm_tools/src/esm_runscripts/prepcompute.py", line 32, in run_job
    config = evaluate(config, "prepcompute", "prepcompute_recipe")
  File "/home/shkjocke/esm/esm_tools/src/esm_runscripts/helpers.py", line 70, in evaluate
    framework_recipe, framework_plugins, config
  File "/home/shkjocke/esm/esm_tools/src/esm_plugin_manager/esm_plugin_manager.py", line 138, in work_through_recipe
    if plugins[workitem]["type"] == "core":
KeyError: 'preprocess'
```

1) It seems that the package problems come from the HLRN anaconda3 environment. Is the new ESM-Tools version not compatible with this environment? 

2) Why does preprocess not work? 

Cheers
Joakim

**Comment by mandresm** [comment]
Hi Joakim, the plugin should install automatically as it is specified in `oifs.yaml`:
https://github.com/esm-tools/esm_tools/blob/ed1d29a4eb167075b40090da35f42033653c2ec3/configs/components/oifs/oifs.yaml#L14-L16

BUT it will only install if you run esm_master to build your model. The good news is that running `esm_master <any_command>-focioifs-<version>` will start by installing those plugins, and you can cancel it whenever it starts the compilation, and then those plugins will be a available for all ESM-Tools commands you run within that environment. You could also install those plugins manually, but I don't think there is a need for that. Please, let me know if that solves the problem.

**Comment by mandresm** [comment]
And by the way, we never really gave support for conda. Now there is a conda `environment.yaml` recently added to `release` (`conda env create -n esm -f environment.yaml; ./install.sh`), and some improvements have been made on that respect, but it is in testing phase. In any case, I think you should be able to run with what you are currently doing with the solution I posted on the message above.

**Comment by joakimkjellsson** [comment]
Hi @mandresm 

I tried the above. 
```bash
cd utils
conda env create -n esm -f environment.yaml
cd .. 
./install.sh
```
did not work (see attached error message). This is the same as I get when I made my own empty env and then ran ./install.sh

But, I switched to "release" branch, then "conda deactivate" (to leave env) and then re-installed esm_tools again:
```bash
rm -rif ~/.local/bin/esm*
rm -rif ~/.local/lib/python3.7/site-packages/esm*
./install.sh
```
Then ran
```bash
esm_master comp-focioifs-2.1 -c
```
which installed plugins. 
And now it works. 

I'm wondering if the "geomar_dev" branch has a different "install.sh" script, which somehow is not compatible with glogin? 
Anyway, it works on "release" which is the important part. 

Thanks for your help!
/Joakim
